Add documentation on how to create executable jar file (Issue #3)

entrylevelcs opened the pull request:
Added instructions for how to create an executable jar file to the README.md which should resolve the [user story from issue #3](https://github.com/MetroCS/imagelab/issues/3).

jody reviewed (changes requested):
This is too specific to the BlueJ IDE.  What if someone is working at the command line or are using JGrasp, Eclipse, or something else?  Do they have to install BlueJ in particular?

It is fine to also provide information about doing so in BlueJ, but for that I believe we're better off linking to the BlueJ-provided info (such as the URL given in the Supporting info), since the instructions could change along with BlueJ.

entrylevelcs commented:
I generalized the instructions for creating the executable jar file.

jody commented on README.md:
```diff
@@ -14,17 +14,21 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
-
+Instructions may vary depending on the IDE but in general the process will be similar to these instructions.
+
+1. Open up the java files from this repository in your desired IDE. These will be found in the **sound, filters and imagelab** folders. Don't forget to also open up the **Run.java** from the base directory of the project. 
```
Why is the "filters" directory included?

jody commented on README.md:
```diff
@@ -14,17 +14,21 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
-7. It should now give you an option to select the main class. The main class for this project is Run, so select Run and click *continue*.
-8. Choose a name for your jar to be saved under. Make sure to save your jar file in base directory of your imagelab repository.
-
+Instructions may vary depending on the IDE but in general the process will be similar to these instructions.
```
Is an IDE required?  Is it not possible to create this JAR using the project's build process? From the command line without using an automated build?

jody commented on README.md:
```diff
@@ -14,17 +14,21 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+5. It should now give you an option to select the main class. The main class for this project is Run, so select Run as the main class.
+6. Choose a name for your jar to be saved under. Make sure to save your jar file in base directory of your imagelab repository.
+
+Some specific examples from popular IDE's can be found here:
+- [Eclipse](https://support.smartbear.com/alertsite/docs/monitors/web/selenium/export-eclipse-java-project-as-runnable-jar.html)
+- [JDeveloper](https://www.albinsblog.com/2014/12/building-executable-jar-file-with.html)
+- [IntelliJ](https://www.jetbrains.com/help/idea/compiling-applications.html#package_into_jar)
+- [BlueJ](https://bluej.org/tutorial/tutorial-v4.pdf)
+- [NetBeans](https://netbeans.org/kb/articles/javase-deploy.html)
```
Having this kind of collection of references is very useful!  Judging from the responses to a poll earlier this semester, VS Code should be included as well.  Given that ImageLab is targeted at those learning to code, it would be useful for DrJava to included too.

jody commented on README.md:
```diff
@@ -14,17 +14,21 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
-
+Instructions may vary depending on the IDE but in general the process will be similar to these instructions.
+
+1. Open up the java files from this repository in your desired IDE. These will be found in the **sound, filters and imagelab** folders. Don't forget to also open up the **Run.java** from the base directory of the project. 
+2. Compile all of the java files.
+3. Navigate your IDE to find its *create jar* functionality.
+4. Set up the dependencies of the jar to use all of the files except for the **Run.java**.
+5. It should now give you an option to select the main class. The main class for this project is Run, so select Run as the main class.
+6. Choose a name for your jar to be saved under. Make sure to save your jar file in base directory of your imagelab repository.
+
```
I would delete all of these "instructions" and replace them with instructions for using the basic Java commands and/or the project's build environment.

Follow that with (an enhanced version of) the set of links to references that you have put together.

entrylevelcs commented:
The project was already set up with the gradle build environment so I changed the instructions to use the preexisting build environment. I also added a couple more references for if the user wanted to create the jar through an IDE.

jody commented on README.md:
```diff
@@ -13,6 +13,21 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+## Creating an executable jar file
+Project is already set up using the gradle build environment so we can just use that to create the jar file.
+1. Open up the CLI and change the directory to the projects root directory.
+2. Execute the command "gradle build"
```
Why "gradle" and not "gradlew" (or "./gradlew")?
What is the name of the jar file and where is it to be found after it is built?
Does the "jar" task not work?  (If not, that would be a good new issue to open.)

entrylevelcs replied:
I wasn't able to get gradlew or jar to work for a while so they were excluded but I figured it out and have added them to the new change.

jody reviewed (commented):
Just a few final questions regarding the "gradle build" command suggestion.

entrylevelcs commented:
Added instructions on how to use "jar" as well as "gradlew" and included resources for further inquiry below both sets of instructions.

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
 
+## Creating an executable jar file
+If there is no automated build environment set up, the jar file can be created using "jar."
+1. Open up the CLI and change the directory to the projects root directory.
+2. Execute the command "javac Run.java" to compile the Run.java file.
+3. Execute the command "javac \*.java" in the package directories to compile/recompile the java files.
```
What are "the package directories"?  If these instructions are so basic as to specify the "javac" command, they shouldn't sometimes presume greater sophistication.

jody replied:
Consider using the `code font` for command literals.

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+2. Execute the command "javac Run.java" to compile the Run.java file.
+3. Execute the command "javac \*.java" in the package directories to compile/recompile the java files.
+4. In the root directory execute the command "echo Main-Class: Run > MANIFEST.MF" to create a manifest file you will use later.
+5. Create the jar using the command "jar cfm (DESIRED_FILENAME.jar) MANIFEST.MF \*.class sound/\*.class imagelab/\*.class filters/\*.class"
```
Why is the `filters` package included in the jar?
Explain how to interpret the notation used, e.g., "(DESIRED_FILENAME.jar)".  Are the parentheses a required part of the command?

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+3. Execute the command "javac \*.java" in the package directories to compile/recompile the java files.
+4. In the root directory execute the command "echo Main-Class: Run > MANIFEST.MF" to create a manifest file you will use later.
+5. Create the jar using the command "jar cfm (DESIRED_FILENAME.jar) MANIFEST.MF \*.class sound/\*.class imagelab/\*.class filters/\*.class"
+6. The new jar file will be named whatever you set it to and it will be located in the projects root directory.
```
Proofread

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+5. Create the jar using the command "jar cfm (DESIRED_FILENAME.jar) MANIFEST.MF \*.class sound/\*.class imagelab/\*.class filters/\*.class"
+6. The new jar file will be named whatever you set it to and it will be located in the projects root directory.
+
+More info on the java commands can be found [here](https://docs.oracle.com/javase/tutorial/deployment/jar/build.html).
```
This seems to be a link to information about creating a JAR file, not a link to info about "java commands" (which might be something more like [JDK Tool Specifications](https://docs.oracle.com/en/java/javase/15/docs/specs/man/index.html)).

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
 A sample set of images are provided in an __"images"__ directory.
 
+## Creating an executable jar file
+If there is no automated build environment set up, the jar file can be created using "jar."
```
I appreciate the inclusion of this information!
It seems like it fits best _after_ the Gradle build instructions. (Even the opening sentence seems to make more sense in that position.)

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+
+However, the project is already set up using the gradle build environment and gradle wrapper so we can just use those to create the jar file.
+1. Open up the CLI and change the directory to the projects root directory.
+2. Execute the command "gradle build" or "./gradlew build"
```
This project intends for the gradle wrapper to be used.  Just delete the phrase "gradle build" or".

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+
+More info on gradle can be found [here](https://docs.gradle.org/current/userguide/userguide.html) and more info on the gradle wrapper can be found [here](https://docs.gradle.org/current/userguide/gradle_wrapper.html).
+
+The executable jar file can also be created straight from an IDE, some specific examples from popular IDE's can be found here:
```
Replace "straight from an IDE, some" with "within an IDE.  Some".

entrylevelcs commented:
Reordered some of the steps, changed some of the instructions to reflect feedback and used `code font` for command literals.

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+2. Execute the command `javac Run.java` to compile the Run.java file.
+3. Execute the command `javac *.java` in the package directories to compile/recompile those package java files. The package directories for this project will be the sound and imagelab directories.
+4. In the root directory execute the command `echo Main-Class: Run > MANIFEST.MF` to create the manifest file you will use for making the jar.
+5. Create the jar using the command `jar cfm DESIRED_FILENAME.jar MANIFEST.MF *.class sound/*.class imagelab/*.class`
+6. The new jar file will retain the name you gave it and will be located in the projects root directory.
```
There's no formatting distinction between DESIRED_FILENAME.jar and MANIFEST.MF but (I believe) they have different meanings: the former is expected to be replaced and the latter is meant to be used literally.

To avoid confusion, I suggest changing `DESIRED_FILENAME.jar` to `imagelab.jar`, which is consistent with the automated build product and leaves no ambiguity.

Line 6 can now read the same as the automated build's line 4:
 ` `  The new jar file will be named "imagelab.jar" and it will be located in the project's root directory.

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
+1. Open up the CLI and change the directory to the projects root directory.
+2. Execute the command `./gradlew build`
+3. This will build/rebuild the jar file using all the required dependencies.
+4. The new jar file will be named "imagelab.jar" and it will be located in the projects root directory.
```
projects => project's

jody commented on README.md:
```diff
@@ -13,6 +13,34 @@ The __filter__ package is generally provided as a folder with source (`.java`) a
 
+## Creating an executable jar file
+The project is already set up using the gradle wrapper so we can just use that to create the jar file.
+1. Open up the CLI and change the directory to the projects root directory.
```
Suggested rewrite:
1. Using the command line interface, change to the project's root directory.

jody reviewed (changes requested):
Looks good overall.  Made a few (final) suggestions for changes.

entrylevelcs commented:
Implemented the suggested changes from the feedback.
